### PR TITLE
synu.walk()

### PR DESCRIPTION
**Pull request by thomasyu888**

### CODE REVIEW

File: synapseutils/copy.py [hunk 1]
```diff
@@ -12,6 +12,8 @@ def copy(syn, entity, destinationId=None, copyWikiPage=True, **kwargs):
     """
     Copies synapse entities including the wikis
 
+    :param syn:             A synapse object: syn = synapseclient.Synapse()
```
Comment (thomasyu888) [inline synapseutils/copy.py]: Specify to log into synapse

File: synapseutils/walk.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import synapseclient
+from synapseclient.entity import is_container
+import os
+
+def walk(syn,synId,newpath=None):
+    """
+    Traverses through the synId specified returning a tuple (directory path, directory names, file names)
```
Comment (thomasyu888) [inline synapseutils/walk.py]: Traverse through the hierarchy of files and folders stored under the synId

File: synapseutils/walk.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        walkedPath = synu.walk(syn, "syn1234")
+
+        for dirpath, dirname, filename in walkedPath:
+            print(dirpath)
+            print(dirname) #All the folders in the directory path
+            print(filename) #All the files in the directory path
+
+    """
+    starting = syn.get(synId,downloadFile=False)
+    if newpath is None and not is_container(starting):
```
Comment (thomasyu888) [inline synapseutils/walk.py]: Return empty generator instead of throw error

File: synapseutils/walk.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+from synapseclient.entity import is_container
+import os
+
+def walk(syn,synId,newpath=None):
+    """
+    Traverses through the synId specified returning a tuple (directory path, directory names, file names)
+
+    :param synId:          A synapse ID of a folder or project
+
+    Example:
```
Comment (thomasyu888) [inline synapseutils/walk.py]: works like os.walk()

File: synapseutils/walk.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import synapseclient
+from synapseclient.entity import is_container
+import os
+
+def walk(syn,synId,newpath=None):
```
Comment (thomasyu888) [inline synapseutils/walk.py]: Write wrapper function and hide newpath = None

File: tests/integration/test_synapseutils.py [hunk 6]
```diff
@@ -385,5 +378,63 @@ def test_copyWiki():
     assert temp.markdown == sub_subwiki.markdown
     assert fourth_header[0] == third_header[0]
 
+def test_walk():
+    walked = []
+    firstfile = utils.make_bogus_data_file()
+    project_entity = syn.store(Project(name=str(uuid.uuid4())))
```
Comment (thomasyu888) [inline tests/integration/test_synapseutils.py]: Move the scheduled clean up

File: tests/integration/test_synapseutils.py [hunk 7]
```diff
@@ -385,5 +378,63 @@ def test_copyWiki():
...
+
+    #Must sort the tuples returned, because order matters for the assert
+    #Folders are returned in a different ordering depending on the name
+    for i in walked:
+        for y in i:
+            y.sort()
+
+    for i in temp:
+        for y in i:
+            y.sort()
```
Comment (thomasyu888) [inline tests/integration/test_synapseutils.py]: for i,j in zip(temp, walked):
   assert set(i)  == set(j)

File: synapseutils/walk.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    return(_helpWalk(syn,synId))
+
+#Helper function to hide the newpath parameter
+def _helpWalk(syn,synId,newpath=None):
+    starting = syn.get(synId,downloadFile=False)
+    #If the first file is not a container, return immediately
+    if newpath is None and not is_container(starting):
+        return
+    elif newpath is None:
+        dirpath = [(starting.name, synId)]
```
Comment (larssono) [inline synapseutils/walk.py]: Should just be tuple.

**Pull request merged**